tortoise/verifying_tortoise.go

```diff
@@ -305,89 +265,158 @@ func (t *turtle) checkBlockAndGetLocalOpinion(
+	}
 
-			logger.With().Info("chose base block",
-				blockID,
-				log.Int("against_count", len(exceptionVectorMap[0])),
-				log.Int("support_count", len(exceptionVectorMap[1])),
-				log.Int("neutral_count", len(exceptionVectorMap[2])))
-
-			metrics.LayerDistanceToBaseBlock.WithLabelValues().Observe(float64(t.Last.Value - layerID.Value))
+	sort.Slice(choices, func(i, j int) bool {
+		ibid := choices[i]
+		jbid := choices[j]
+		// prioritize good blocks
+		_, iexist := t.GoodBlocksIndex[ibid]
+		_, jexist := t.GoodBlocksIndex[jbid]
+		if iexist != jexist {
+			return iexist
+		}
+		// prioritize blocks with less disagreements to a local opinion
+		if disagreements[ibid] != disagreements[jbid] {
+			return disagreements[ibid].After(disagreements[jbid])
+		}
+		// priortize blocks from higher layers
+		if t.BlockLayer[ibid] != t.BlockLayer[jbid] {
+			return t.BlockLayer[ibid].After(t.BlockLayer[jbid])
+		}
+		// otherwise just sort determistically
+		return ibid.Compare(jbid)
+	})
 
-			return blockID, [][]types.BlockID{
-				blockMapToArray(exceptionVectorMap[0]),
-				blockMapToArray(exceptionVectorMap[1]),
-				blockMapToArray(exceptionVectorMap[2]),
-			}, nil
+	for _, bid := range choices {
+		lid := t.BlockLayer[bid]
+		exceptions, err := t.calculateExceptions(tctx, lid, t.BlockOpinionsByLayer[lid][bid])
+		if err != nil {
+			logger.With().Warning("error calculating vote exceptions for block", bid, log.Err(err))
+			continue
 		}
+		logger.With().Info("chose base block",
+			bid,
+			log.Int("against_count", len(exceptions[0])),
+			log.Int("support_count", len(exceptions[1])),
+			log.Int("neutral_count", len(exceptions[2])))
+
+		metrics.LayerDistanceToBaseBlock.WithLabelValues().Observe(float64(t.Last.Value - lid.Value))
+
+		return bid, [][]types.BlockID{
+			blockMapToArray(exceptions[0]),
+			blockMapToArray(exceptions[1]),
+			blockMapToArray(exceptions[2]),
+		}, nil
 	}
 
 	// TODO: special error encoding when exceeding exception list size
-	return types.BlockID{0}, nil, errNoBaseBlockFound
+	return types.BlockID{}, nil, errNoBaseBlockFound
+}
+
+// firstDisagreement returns first layer where local opinion is different from blocks opinion within sliding window.
+func (t *turtle) firstDisagreement(ctx *tcontext, blid types.LayerID, bid types.BlockID, disagreements map[types.BlockID]types.LayerID) (types.LayerID, error) {
+	var (
+		opinions = t.BlockOpinionsByLayer[blid][bid]
+		from     = t.LastEvicted.Add(1)
+		// using it as a mark that the votes for block are completely consistent
+		// with a local opinion. so if it two blocks have consistent histories select block
+		// from a higher layer as it is more consistent.
+		consistent = t.Last
+	)
+
+	// if the block is good we know that there are no exceptions that precede base block.
+	// in such case current block will not fix any previous disagreements of the base block.
+	// or if the base block completely agrees with local opinion compute disagreements after base block.
```

Comment:
i don't understand this comment.
if we already considered `bid` as a good block, that means
- we once considered base block of `bid` as a good block.
- our local opinion agrees with all exceptions that `bid` has.

so what do you mean by `current block will not fix any previous disagreements of the base block`

Reply:
we don't consider implicit votes against the blocks when we are considering if block is good, so block might be missing some votes, which is exactly the problem i am addressing in this pr

Reply:
when i read your code here, i thought the problem is that we currently don't generate AgainstDiff. so even if we find a base block that's good, it's not really good.

your comment only make sense when we don't generate AgainstDiff.

Reply:
AgainstDiff will be populated with votes that are different from base block votes.

it doesn't change implicit votes, e.g. votes for blocks that we are not aware of, for example a gap vote in layer 12 with base block at layer 10 will implicitly vote against blocks at layer 11. so even if block 10 is good, we still need to check that implicit votes for layer 11 match our local opinion

Reply:
> AgainstDiff will be populated with votes that are different from base block votes.

we still don't add AgainstDiff to the block. 
https://github.com/spacemeshos/go-spacemesh/blob/6a93e4781ccb5091a0b4dcaad811f4ebb8aa0dc8/tortoise/verifying_tortoise.go#L506

what am i missing?

Reply:
yes we are not adding, i guess i don't understand what we are discussing.

> your comment only make sense when we don't generate AgainstDiff.

why is that? if there was a disagreement because the base block didn't vote for some layer, and a new block is voting for that layer - such block will not be marked good, but it may fix base block disagreements, so we need to do more work. however if we know that the block is good - then all we need to check if later votes are consistent with local opinion, including implicit votes.

Reply:
let me start by saying, the bug is weird in the first place. if we always choose a good block in layer N, and the definition of a good block is that 
- its base block in layer M was a good block according to the local opinions
- the block had no disagreement with our local opinions between layer M and N

then why do we end up with exceptions older than the base block? the only reason i can think of is that we are not adding AgainstDiff to blocks and so the expanded votes based on base block is incorrect. and somehow in the code we end up with disagreements even btwn good blocks in different layers.

so when you are comments are talking about disagreement even with the base block (which we locally considered a good block), i immediately thought about the lack of AgainstDiff in the diff list.

> it doesn't change implicit votes, e.g. votes for blocks that we are not aware of, for example a gap vote in layer 12 with base block at layer 10 will implicitly vote against blocks at layer 11. so even if block 10 is good, we still need to check that implicit votes for layer 11 match our local opinion

i still don't understand what you mean by "votes for blocks that we are not aware of". 

let's for a moment disregard what the code does and discuss what should be in the data.
when we are creating votes for block `12B` in layer 12. we chose block `10A` in layer 10 to be base block. 
two scenarios:
- layer 11 has not terminated yet. then block `12B` should have have 
```
base block: 10A
forDiff: all block IDs in layer 10
againstDiff: empty
neutralDiff: all block IDs in layer 11
```
- layer 11 terminated with failure. then block `12B` should have have 
```
base block: 10A
forDiff: all block IDs in layer 10
againstDiff: all block IDs in layer 11
neutralDiff: empty
```
these are the Block data over the wire.
implicit against votes should only happen when both block `12B` and block `10A` voted against a block in a past layer, say `9Z`. over the wire, block `10A` should still have `9Z` in its AgainstDiff, but `12B` will not, because it agrees with `10A`

so i think what the vote gap you are referring to is in the second case when hare terminated layer 11 with failure, block `12B` actually looked like
```
base block: 10A
forDiff: all block IDs in layer 10
againstDiff: empty
neutralDiff: empty
```
i think this is where the confusion come from during our discussion. the `implicit against` is too confusing and mind-twisting especially with the recursive nature of the base block logic.

Reply:
thanks for the detailed explanation, it all makes sense except that i am not sure that the conditions that you specified are exhaustive.

1. layer 11 terminated with failure. then block 12B should have have
2.

Reply:
thank you for the detailed explanation, it all makes sense but i don't think that the specified conditions are exhaustive.

```
base block: 10A
forDiff: all block IDs in layer 10
againstDiff: all block IDs in layer 11
neutralDiff: empty
```

in this example, you assume that we know all block ids that were in the layer 11. f the node doesn't know any or knows only 1 out of 10, how this encoding will change?

this is why i don't think that we can get rid of "implicit" votes against unknown blocks.

so for me, it is a question of when do we consider these "implicit" votes. in this implementation, we consider them only when we count votes for blocks, and with this change, during base block selection. this is probably a mistake because...

> its base block in layer M was a good block according to the local opinions
   the block had no disagreement with our local opinions between layer M and N

of this difference with the implementation, as we don't consider implicit votes when goodness is determined it means that the block can have a disagreement with local opinion but still can be marked good. we only check explicitly that provided exceptions are consistent with local opinion.

Reply:
> in this example, you assume that we know all block ids that were in the layer 11. f the node doesn't know any or knows only 1 out of 10, how this encoding will change?

@dshulyak 
the encoding won't change. you still add that 1 block you know about in the `againstDiff`. this means when the node is voting about layer 11, either the hare terminate layer 11 with 1 block (if we are voting within hdist) or the votes you see from the blocks only have 1 block for layer 11. the rest of 9 blocks must have appeared late in the network.

the implicit against is worse. you have no idea what the node saw for layer 11. it unequivocally votes against all blocks in a layer. doesn't matter whether the node actually saw 1 or 2 or all 10 blocks. i don't see this as an advantage for implicit votes. i see this as an disadvantage because we lose information.

Reply:
but of course when you know there are 10 blocks in layer 11, and the vote has only one block in its `againstDiff`, you can still explicitly vote against the rest of the 9 blocks when we expand the voting vector in memory. the point is, at the time of voting, we know exactly what the node saw.

Reply:
> of this difference with the implementation, as we don't consider implicit votes when goodness is determined it means that the block can have a disagreement with local opinion but still can be marked good. we only check explicitly that provided exceptions are consistent with local opinion.

are you sure this is not simply because we don't encode add `againstDiff` now? if we add `againstDiff` to the current implementation, a good block should satisfy this property

> its base block in layer M was a good block according to the local opinions
the block had no disagreement with our local opinions between layer M and N

and in this case, if a block B had used a good block as base block, and block B is also determined as a good block, there should not be any disagreement btwn block B and local opinions.

Reply:
> the encoding won't change.

it will change as you noted below

> but of course when you know there are 10 blocks in layer 11, and the vote has only one block in its againstDiff, you can still explicitly vote against the rest of the 9 blocks when we expand the voting vector in memory. the point is, at the time of voting, we know exactly what the node saw.

this is was my point in a previous reply https://github.com/spacemeshos/go-spacemesh/pull/2928#discussion_r748793531 . but this is not explicit voting, since you don't add those blocks as exceptions. 

and as i mentioned this is just a question of when do you count those vote, in the current implementation we count them later, which is probably a mistake that follows from the imprecise implementation of good block.

> are you sure this is not simply because we don't encode add againstDiff now? if we add againstDiff to the current implementation, a good block should satisfy this property

i am sure, we still need to consider implicit votes as you noted above